mesonbuild/modules/features/feature.py

```diff
@@ -0,0 +1,356 @@
+            val = getattr(self, attr)
+            if not val:
+                val = ''
+            else:
+                val = str(val)
+            ret[attr] = val
+        return ret
+
+class KwargConfilctAttr(KwargInfo):
+    def __init__(self, func_name: str, opt_name: str, default: Any = None):
+        types = (
+            NoneType, str, ContainerTypeInfo(dict, str),
+            ContainerTypeInfo(list, (dict, str))
+        )
+        super().__init__(
+            opt_name, types,
+            convertor = lambda values: self.convert(
+                func_name, opt_name, values
+            ),
+            default = default
+        )
+
+    @staticmethod
+    def convert(func_name:str, opt_name: str, values: 'IMPLIED_ATTR',
+                ) -> Union[None, List[ConflictAttr]]:
+        if values is None:
+            return None
+        ret: List[ConflictAttr] = []
+        values = [values] if isinstance(values, (str, dict)) else values
+        accepted_keys = ('val', 'match', 'mfilter', 'mjoin')
+        for edict in values:
+            if isinstance(edict, str):
+                if edict:
+                    ret.append(ConflictAttr(val=edict))
+                continue
+            if not isinstance(edict, dict):
+                # It shouldn't happen
+                # TODO: need exception here
+                continue
+            unknown_keys = [k for k in edict.keys() if k not in accepted_keys]
+            if unknown_keys:
+                raise MesonException(
+                    f'{func_name}: unknown keys {unknown_keys} in '
+                    f'option {opt_name}'
+                )
+            val = edict.get('val')
+            if val is None:
+                raise MesonException(
+                    f'{func_name}: option "{opt_name}" requires '
+                    f'a dictionary with key "val" to be set'
+                )
+            implattr = ConflictAttr(val=val, mjoin=edict.get('mjoin', ''))
+            for cattr in ('match', 'mfilter'):
+                cval = edict.get(cattr)
+                if not cval:
+                    continue
+                try:
+                    ccval = re.compile(cval)
+                except Exception as e:
+                    raise MesonException(
+                        '{func_name}: unable to '
+                        f'compile the regex in option "{opt_name}"\n'
+                        f'"{cattr}:{cval}" -> {str(e)}'
+                    )
+                setattr(implattr, cattr, ccval)
+            ret.append(implattr)
+        return ret
+
+if TYPE_CHECKING:
+    IMPLIED_ATTR = Union[
+        None, str, Dict[str, str], List[
+            Union[str, Dict[str, str]]
+        ]
+    ]
+    class FeatureKwArgs(TypedDict):
+        #implies: Optional[List['FeatureObject']]
+        implies: NotRequired[List[Any]]
+        group: NotRequired[List[str]]
+        detect: NotRequired[List[ConflictAttr]]
+        args: NotRequired[List[ConflictAttr]]
```

Comment:
I would stick the typing logic up at the top, still.

ConflictAttr could be quoted in an `if T.TYPE_CHECKING:` block to cause it to parse as a forward reference, it doesn't *need* to be defined before.